src/scanpy/preprocessing/_pca/_dask_sparse.py

```diff
@@ -0,0 +1,185 @@
+
+        return da.map_blocks(
+            transform_block,
+            x,
+            mean_=self.mean_,
+            components_=self.components_,
+            chunks=(x.chunks[0], self.n_components_),
+            meta=np.zeros([0], dtype=x.dtype),
+            dtype=x.dtype,
+        )
+
+
+@overload
+def _cov_sparse_dask(
+    x: DaskArray, *, return_gram: Literal[False] = False, dtype: DTypeLike | None = None
+) -> tuple[NDArray[np.floating], NDArray[np.floating]]: ...
+@overload
+def _cov_sparse_dask(
+    x: DaskArray, *, return_gram: Literal[True], dtype: DTypeLike | None = None
+) -> tuple[NDArray[np.floating], NDArray[np.floating], NDArray[np.floating]]: ...
+def _cov_sparse_dask(
+    x: DaskArray, *, return_gram: bool = False, dtype: DTypeLike | None = None
+) -> (
+    tuple[NDArray[np.floating], NDArray[np.floating], NDArray[np.floating]]
+    | tuple[NDArray[np.floating], NDArray[np.floating]]
+):
+    """\
+    Computes the covariance matrix and row/col means of matrix `x`.
+
+    Parameters
+    ----------
+
+    x
+        A sparse matrix
+    return_gram
+        If `True`, the gram matrix will be returned and a copy will be created
+        to store the results of the covariance,
+        while if `False`, the local gram matrix result will be overwritten.
+        (only used for unit testing at the moment)
+    dtype
+        The data type of the result (excluding the means)
+
+    Returns
+    -------
+
+    :math:`\\cov(X, X)`
+        The covariance matrix of `x` in the form :math:`\\cov(X, X) = \\E(XX) - \\E(X)\\E(X)`.
+    :math:`\\gram(X, X)`
+        When return_gram is `True`, the gram matrix of `x` in the form :math:`\\frac{1}{n} X.T \\dot X`.
+    :math:`\\mean(X)`
+        The row means of `x`.
+    """
+    if TYPE_CHECKING:
+        import dask.array.core as da
+        import dask.base as dask
+    else:
+        import dask
+        import dask.array as da
+
+    if dtype is None:
+        dtype = np.float64 if np.issubdtype(x.dtype, np.integer) else x.dtype
+    else:
+        dtype = np.dtype(dtype)
+
+    def gram_block(x_part: CSMatrix):
+        gram_matrix: CSMatrix = x_part.T @ x_part
+        return gram_matrix.toarray()[None, ...]  # need new axis for summing
+
+    gram_matrix_dask: DaskArray = da.map_blocks(
+        gram_block,
+        x,
+        new_axis=(1,),
+        chunks=((1,) * x.blocks.size, (x.shape[1],), (x.shape[1],)),
+        meta=np.array([], dtype=x.dtype),
+        dtype=x.dtype,
+    ).sum(axis=0)
+    mean_x_dask, _ = _get_mean_var(x)
+    gram_matrix, mean_x = cast(
+        tuple[NDArray, NDArray[np.float64]],
+        dask.compute(gram_matrix_dask, mean_x_dask),
+    )
+    gram_matrix = gram_matrix.astype(dtype)
+    gram_matrix /= x.shape[0]
+
+    cov_result = gram_matrix.copy() if return_gram else gram_matrix
+    cov_result -= mean_x[:, None] @ mean_x[None, :]
+
+    if return_gram:
+        return cov_result, gram_matrix, mean_x
+    return cov_result, mean_x
```

Comment:
I still don't get why this is here if the only place we actually use this is in the tests.  We just don't need to test it?

Reply:
https://github.com/scverse/scanpy/pull/3263#discussion_r1783023544

I don’t feel super strongly about this, but I feel like having a checkpoint is better than having none.

Reply:
Ok, We can leave it.